### ISSUE TITLE
[analyzer] redundant nullable variable

### ISSUE THREAD

**Issue by asashour** [dart-lang/sdk#48723]
For a code like:

```dart
void f(int x) {
  int? i = x;
}
```

`i` is always not nullable.

Would it be helpful to have a hint about this redundancy?

**Comment by eernstg** [comment]
It is already being handled in the following sense: When a variable has a nullable declared type and it has an initializing expression whose type is the corresponding non-nullable type, then the variable is immediately promoted. Spec language [here](https://github.com/dart-lang/language/blob/master/accepted/2.12/nnbd/feature-specification.md#local-variables-and-inference).

This means that subsequent code will be able to use `i` with the non-nullable type (such that `i.isEven` is not a compile-time error), as long as it isn't demoted (as it would be, e.g., after `i = null`). So it could be argued that the situation isn't much of a problem: The variable has the "nice" type anyway, and we're assuming that the developer has the intention that null could be assigned to `i` at some point.

It could be argued that there should be a diagnostic message to suggest that a local variable `x` with declared type `T` should should be changed to have declared type `S`, where `S` is the least subtype of `T` which will not cause any expression in the entire scope of `x` to become a compile-time error, if any such type exists.

However, that could be a bad idea: It is often recommended that local variables are declared with a somewhat abstract type (for instance, in Java code, it's commonly considered better to use `List<...>` than `ArrayList<...>` as the type of a local variable&mdash;and in Dart we've hidden the concrete list implementation types so well that there is no debate about that question at all).

In summary, it would be possible to flag situations like this, but there are multiple reasons why it might not be needed so urgently.

**Comment by bwilkerson** [comment]
I mostly agree with that, though I have to wonder whether nullability might be slightly different than the more general case from the user's perspective.

If we were to add something we would probably want to make it a lint so that interested users could opt in to this kind of information, but users that don't care won't be bothered.
